### ISSUE TITLE
Consideration for Shadow DOM?

### ISSUE THREAD

**Issue by bradleyayers** [typestyle/typestyle#92]
Has any consideration been given to compatibility with Shadow DOM?

The fundamental requirement is being able to choose where the `<style>` element is added to the DOM (so that you target elements in a Shadow DOM by adding the `<style>` to that same Shadow DOM).

**Comment by basarat** [comment]
> Has any consideration been given to compatibility with Shadow DOM?

Not yet...but:

> The fundamental requirement is being able to choose where the <style> element is added to the DOM 

Should already work. We don't call `getTag` till first async completes (request animation frame or set timeout). `getTag` creates the tag lazily : https://github.com/typestyle/typestyle/blob/master/src/index.ts#L44-L50 

So if you call `setTag` with some other tag : https://github.com/typestyle/typestyle/blob/master/src/index.ts#L51-L59 it should just work :rose:

Please leave a comment if it doesn't work out 👍

**Comment by bradleyayers** [comment]
Looking at the code, the problem would appear to be that there's a single stylesheet containing all the styles, which wouldn't be compatible with Shadow DOM as you need at least one-stylesheet-per-shadow-dom if you want to be able to style each Shadow DOM separately.

I think the solution would need to take the form of a factory function that takes a destination DOM element (as the `<style>` container) yields a `style` function that behaves as it does today.

A consumer that wants to put styles into a Shadow DOM, they'd need to do something like:

```js
import { createStyle } from 'typestyle';
const style = createStyle({ container: document.getElementById('myShadowHost').shadowRoot });
const button = style({ borderRadius: 3 });

// …
```

[Fela](http://fela.js.org/) has flexility that I think facilitates this kind of approach.

**Comment by basarat** [comment]
That is a pretty different design than what we have. Having different stylesheets (instead of a singleton) would mean that none of the deduping works. The different design is indeed having *different instances of typestyle* (and internally freestyle) running in parallel. 

Why can't the global stylesheet be used to contain the styles for the components? If you were e.g. using css-modules how would you handle it?

**Comment by bradleyayers** [comment]
The global stylesheet can't be used because Shadow DOM creates a boundary through which CSS does not pass (apart from a few specific properties). This means that in order to style content within a Shadow DOM, you need a `<style>` *within the Shadow DOM itself*.

If you're using CSS modules you can't use the webpack `style-loader` (since it too puts the stylesheet in the main document), instead you use the output of `css-loader` directly and have something like:

```js
import styles from 'css-loader!styles.css';
import { Component } from 'skatejs';

class MyWebComponent extends Component {
  static render() {
    return (
      <div>
        <style>{styles.toString()}</style>
        <div className={styles.locals.button>Click me</div>
      </div>
    );
  }
}
```

In practice though you wouldn't bother with CSS modules here, since the main benefit it brings is scoping in the global namespace (which is already solved by Shadow DOM), so in practice you'd probably just do:

```js
import { Component } from 'skatejs';

class MyWebComponent extends Component {
  static render() {
    return (
      <div>
        <style>
        .button { background: red; }
        <style>
        <div className="button">Click me</div>
      </div>
    );
  }
}
```

TypeStyle offers something different to CSS modules, because it aspires to offer a type-safe API for declaring CSS — something that Shadow DOM makes no improvement towards. This means that while CSS modules are not particularly useful in a Shadow DOM world, TypeStyle still would/could be.

**Comment by basarat** [comment]
Thanks for your expert sharing of knowledge and sorry for inexperience in this area. 

The main original motivation I had for typestyle over freestyle was actually fold : safety + zero-config. 

I'd like to think about this more. Definitely should be kept open till we can think of an easy solution that doesn't compromise the zero-config for the normal use case :rose: 

I am currently busy with http://basarat.com/eze :heart:

**Comment by notoriousb1t** [comment]
Some off the cuff thoughts..  If we restructured the internals of typestyle to be a class, and then piped all of the exports to an instance of it, that would maintain the zero config while making the general functionality transferable.  Then we could expose a function that returns a new instance of that class used to target additionally style tags.

```ts
const defaultTypeStyle = new TypeStyle();

class TypeStyle {
    style(/* */);
}

export style = (...props) => {
    return defaultTypeStyle.style(...props);
}

export const typestyle(target: Element|string): TypeStyle {
   const instance = new TypeStyle();
   instance.setStylesTarget(target);
   return instance;
}
```

I can branch to see what that would look like

**Comment by basarat** [comment]
@notoriousb1t Sounds great! :rose:  

A tip:

```ts
class TypeStyle {
    style = (/* */) => { /* */ };
}

export const style = defaultTypeStyle.style;
```
Also, I am sure you will do the following, just wanted to be helpful : 

```ts
// Just the type annotation
export function typestyle(target:{ textContent: string | null }): TypeStyle {
```
:heart:

**Comment by notoriousb1t** [comment]
Will do 😄 .

@basarat Do you want me to go with a class or use a factory?

**comparison**
_class_: https://github.com/just-animate/just-animate/blob/c41d9fe0d2228221309a01e3af75395725648391/src/plugins/core/Dispatcher.ts

_factory_: https://github.com/just-animate/just-animate/blob/04d2221bb70dfd6848793bb7d01a0cf369b56fc3/src/plugins/core/Dispatcher.ts

**Comment by basarat** [comment]
a class in `/internal/core`. I like classes https://basarat.gitbooks.io/typescript/content/docs/tips/classesAreUseful.html So does free-style apparently : https://github.com/blakeembrey/free-style/blob/f6ca5630ff73f22cf6bb5f12fccd7b33209b710e/src/free-style.ts#L491-L521 :rose:

**Comment by basarat** [comment]
Aka happy with your original :heart:

```ts
export function typestyle(target:{ textContent: string | null }): TypeStyle {
   const instance = new TypeStyle();
   instance.setStylesTarget(target);
   return instance;
}
```

**Comment by notoriousb1t** [comment]
There is some nuance to choosing one or the other (encapsulation vs performance), but classes sound good to me 😄.  I don't think there is going to be a lot of motivation for people to mess with the internals.  Classes are also way easier to read, so I am good for that reason as well 👍

**Comment by basarat** [comment]
Because of your excellent code / thought contributions to TypeStyle (it wouldn't be what it is without you :heart:) thought I'd share my thoughts: 

```ts
var iterations = 10000000; 
var foo = 0;  var Bar = class { /* some local */ constructor(){this.bar = 123;} inc(){foo++; this.bar++}}; console.time('class');for(let i = 0; i<iterations ; i++){new Bar().inc()}console.log(foo);console.timeEnd('class');
var foo = 0;  function createBar(){ /* some local */ let bar = 123; return {bar,inc:function(){foo++,bar++}}}; console.time('fact');for(let i = 0; i<iterations; i++){createBar().inc()}console.log(foo);console.timeEnd('fact');
```

So no need to feel guilty about classes no more :rose: 

![image](https://cloud.githubusercontent.com/assets/874898/21730061/098ffcba-d4a2-11e6-97c1-ceb0cafab48c.png)

Of course that is native es6 classes but just wanted to show that its generally easier to just write code that's easy to maintain as browsers can do pretty fancy and sometimes somewhat random stuff :heart:

And every microbenchmark can be forced to skew to ones opinion 😉

**Comment by notoriousb1t** [comment]
@basarat I created a branch with proposed changes: https://github.com/typestyle/typestyle/tree/shadow-branch

I had to bind the zero config functions to the defaultInstance in order for them to work right.  I created a bindToInstance function that does the bind manually and lies to the typescript compiler so the typings return properly. 

 It feels weird to downcast to any and then upcast to the input type, but it keeps the typings correct on the export without having to defined them a second time.

**Comment by notoriousb1t** [comment]
I also made it possible for TypeStyle to function without a style target to make it easier to use with Angular 2 component styles: https://angular.io/docs/ts/latest/guide/component-styles.html

**Comment by basarat** [comment]
Thanks @notoriousb1t, released with `0.23.0` :heart: @bradleyayers  give it a go please. There is a new `createTypeStyle` function exported. Will add docs once you okay it as fit for usage :rose:

**Comment by basarat** [comment]
@bradleyayers did you get a chance to try this. If not its okay. This refactoring was worth it :rose: 

If you did and it didn't work, leave a comment :heart:

**Comment by bradleyayers** [comment]
Sorry, a bit late to the party here. I haven't got an immediate need for this, but will definitely keep it in mind and report back when I use it 👍 .

**Comment by greyepoxy** [comment]
Hey just wanted to let you know that I am using `createTypeStyle` to allow for my UI code to be tested using https://github.com/jamesshore/quixote/. It uses an iframe  in order to test CSS layout so needed my styles to be created in the frame.

I can create a new issue if needed but there are several places in the code that just assume the default instance of typestyle and do not support using an instance created with `createTypeStyle`.
1. https://github.com/typestyle/csstips/blob/master/src/normalize.ts
1. https://github.com/typestyle/csstips/blob/master/src/page.ts

and when you are dependent on using an instance like this it would be nice to be able to get at the default one https://github.com/typestyle/typestyle/blob/master/src/index.ts#L17 for when I am not testing and just trying to render my full app.
